### ISSUE TITLE
javascript light wallet

### ISSUE THREAD

**Issue by zack-bitcoin** [aeternity/testnet#10]
we need javascript libraries for downloading the headers from the blocks, and finding the longest chain of headers.
For now only the experimental branch has headers for the blocks.

**Comment by stefek99** [comment]
OK.

I specialise in JavaScript, this is something I could do.

Will need some guidance.

**Comment by zack-bitcoin** [comment]
There is already some example javascript code. I haven't been maintaining it recently, but some parts still work. 
most important is rpc.js, which is used for accessing a node's api.
In src/web and src/external_web

I need to add some more commands to the api for downloading just the header.
You can see how headers are formatted in /src/consensus/tree/block.erl

For now if you can make a piece of javascript code that can verify a hash that was computed on a node, that you would be great.

Here are some example hashes: 2> testnet_hasher:doit(<<1,2,3,4>>).
<<254,228,110,28,173,97,24,147,123,120,21,176>>
4> testnet_hasher:doit(<<"bob">>).
<<70,158,4,119,63,43,188,208,57,190,132,38>>
5> testnet_hasher:doit(<<>>).
<<182,19,103,154,8,20,217,236,119,47,149,215>>

Here is an example of the hash converted into a json object:
12> packer:pack({abc, testnet_hasher:doit(<<>>)}).
<<"[\"abc\",\"thNnmggU2ex3L5XX\"]">>
13> io:fwrite(packer:pack({abc, testnet_hasher:doit(<<>>)})).
["abc","thNnmggU2ex3L5XX"]

This is a sha256 hash, I think there are some bitcoin javascript libraries to compute this.

**Comment by stefek99** [comment]
First what I'll do is to find reliable, audited crypto library to calculate SHA256 and we will start from there.

> For now if you can make a piece of javascript code that can verify a hash that was computed on a node, that you would be great.

Is it about checking if the hash is the same?

**Comment by stefek99** [comment]
I need to better understand what commands / options are available via API.

For now: http://plnkr.co/edit/Rqc7Yv45SUq0DuKCMQAA?p=preview

```
<!doctype html>
<html>
  <head>

    <title>SHA256 demo</title>

    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
    <script src='http://ajax.googleapis.com/ajax/libs/angularjs/1.5.7/angular.min.js'></script>

    <!-- using http://stackoverflow.com/questions/18338890/are-there-any-sha-256-javascript-implementations-that-are-generally-considered-t -->
    <!-- TODO: include only relevant functions -->
    <!-- TOOD: to the same trick with 'integrity' as Bootstrap does -->
    <script src='https://bitwiseshiftleft.github.io/sjcl/sjcl.js'></script>
    <script>
      var app = angular.module("app", []);

      app.controller("ctrl", function($scope) {
      
        $scope.calculateHash = function() {
          var out = sjcl.hash.sha256.hash($scope.input);
          $scope.hash = hash = sjcl.codec.hex.fromBits(out);
        };


      });

    </script>
  </head>
  <body ng-app="app" ng-controller="ctrl" class="container">
    <h1>Calculate SHA256 hash</h1>
    
    <form>
      <textarea class="form-control" ng-model="input" ng-change="calculateHash()" rows=10></textarea>

      <br>

      <input class="form-control" ng-model="hash" disabled>
    </form>


  </body>
</html>
```

**Comment by Leonyxtested** [comment]
Esta interesante lo que has hecho,escribes lo que quieras y lo combierte a SHA256, en la pagina de Erlang puede que encuentres algo http://erlang.org/doc/search/
 En readme.md vienen unos comandos basicos para el nodo,quizas tu necesites algo más tecnico,no lo se.

**Comment by stefek99** [comment]
@Leonyxtested 

> I need to better understand what commands / options are available via API

`testnet_hasher:doit(<<1,2,3,4>>).`

` packer:pack({abc, testnet_hasher:doit(<<>>)}).`

` io:fwrite(packer:pack({abc, testnet_hasher:doit(<<>>)})).`

@zack-bitcoin all this project is not a job for a single dev, you definitely need some reinforcements. Some of my friend from university (I know, I know) are pretty badass at Erlang. Can reach out but... It's me not being proactive, it's me not stepping on your toes.

**Comment by Leonyxtested** [comment]
Si no me equivo lo que va antes de los dos puntos es el documento y luego el comando que ejecuta,si no me equivo es:

testnet_hasher:doit(<<1,2,3,4>>). Una lista que define el tamaño del hash,doit,hace referencia ha otro comando.

packer:pack({abc, testnet_hasher:doit(<<>>)}). Codifica la lista llamada "abc",es el ejemplo que esta usando en el test,que es el comando anterior.

io:fwrite(packer:pack({abc, testnet_hasher:doit(<<>>)})). Muestra en pantalla como quedaria todo lo anterior,me parece que en binario o en base64,dependiendo como se haya hecho.

@zack-bitcoin te podra responder mejor

**Comment by stefek99** [comment]
It's not rocket science, it's kind of obvious what `io:fwrite` does :)

However, I'm not sure if `packer:pack` is an utility created by you, or some built-in Erlang feature.

I suggested - https://github.com/aeternity/testnet/issues/23 - so that there is no need to worry about backend.

I also created super simple SHA256 demo - https://github.com/aeternity/testnet/issues/10#issuecomment-292820474

> For now if you can make a piece of javascript code that can verify a hash that was computed on a node, that you would be great.

Need more info, be more specific. Suggested Skype / Hangout at 19:06 my local time.

**Comment by Leonyxtested** [comment]
Yo no lo sabia hasta que lo lei en el manual,siento haber sido obvio.
> It's not rocket science, it's kind of obvious what io:fwrite does :)

`packer:pack` es una utilidad creada,dentro de esta esta `jiffy:encode` que es una utilidad creada por otra persona.

La información especifica te la puede dar el programador,ojalá te pudiera concretar más.

**Comment by zack-bitcoin** [comment]
Yes, all we need is to take the hash of a binary and verify that it matches what we expect.
I am not up for a call today, sorry. Ive got a headache. Maybe tomorrow.
You can see the api commands.
The ones available externally are here: https://github.com/aeternity/testnet/blob/master/src/networking/handler.erl
The ones that are only available on the same node are here: https://github.com/aeternity/testnet/blob/master/src/networking/internal_handler.erl

Here is the line of javascript code that accesses your balance from the internal handler: https://github.com/aeternity/testnet/blob/master/src/web/balance.js#L20

Hiring more people might help. Applicants should feel free to contact us. zack@aeternity.com is my email to start the process.

leonyxtested is right.
packer:pack is a wrapper that slightly modifies the jiffy library that someone else wrote.
jiffy is downloaded and compiled by rebar. see rebar.config for all the libraries that Aeternity depends on.

Cool demo, I will try it soon.

**Comment by stefek99** [comment]
> You can see the api commands.
> The ones available externally are here: https://github.com/aeternity/testnet/blob/master/src/networking/handler.erl

Sorry, it would be easier if you could say: "GET request to http://localhost/balance"

RESTful conventions...
I can investigate what `variable_get` is doing and reverse engineer the URLs...

(not tonight though, too tired)

**Comment by zack-bitcoin** [comment]
@stefek99 I tried out the javascript hasher, It looks like we are close, but it isn't the same yet.

I converted one of the erlang hash commands to hex for easier comparison:
1> testnet_hasher:bin_to_hex(testnet_hasher:doit(<<>>)).
["B","6","1","3","6","7","9","A","0","8","1","4","D","9",
 "E","C","7","7","2","F","9","5","D","7"]

**Comment by zack-bitcoin** [comment]
I agree it would be nice to just do localhost/balance, but I wanted the api to accept and produce full JSON objects. I wasn't sure how to embed a json object into the URL that way.
The info is _not_ embedded into the url.
Have a look here how I access the same API from bash using the curl command:
https://github.com/aeternity/testnet/blob/master/channel_test.sh

**Comment by stefek99** [comment]
The CURL just makes web requests - doable purely in JavaScript.

* https://en.wikipedia.org/wiki/Same-origin_policy
* https://en.wikipedia.org/wiki/Cross-origin_resource_sharing

```
Access-Control-Allow-Origin: http://www.example.com
Access-Control-Allow-Methods: GET, POST
```



How to run ```testnet_hasher:bin_to_hex(testnet_hasher:doit(<<>>)).```?

Here is what I try to do:

```
sh start.sh
From https://github.com/aeternity/testnet
 * branch            master     -> FETCH_HEAD
Already up-to-date.
==> pink_hash (get-deps)
==> dump (get-deps)
==> trie (get-deps)
==> chalang (get-deps)
==> jiffy (get-deps)
==> cowlib (get-deps)
==> ranch (get-deps)
==> cowboy (get-deps)
==> testnet (get-deps)
==> pink_hash (compile)
==> dump (compile)
==> trie (compile)
==> chalang (compile)
==> jiffy (compile)
==> cowlib (compile)
==> ranch (compile)
==> cowboy (compile)
==> testnet (compile)
GO TO THIS WEBSITE -------> http://localhost:8041/login.html
Erlang/OTP 19 [erts-8.3] [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Eshell V8.3  (abort with ^G)
1> start keys
start toptx pool started
start server
testnet_hasher:bin_to_hex(testnet_hasher:doit(<<>>)).
** exception error: undefined function testnet_hasher:bin_to_hex/1
2>
```

**Comment by zack-bitcoin** [comment]
I just wrote bin_to_hex. It isn't in Aeternity.
Ill push it now for your convenience.

**Comment by stefek99** [comment]
Please - never do anything for me.

You can do something for the project, humanity, planet Earth - but never for me.

I'm asking basic questions - and if I don't know how to run a command there is a non-zero chance that someone else may also not know it.

**Comment by GroundedSage** [comment]
I am also willing to chip in where I can. I'm much less experience in vanilla js and focus more on cljs. But wherever I can help I will. Even if it's writing docs.

**Comment by stefek99** [comment]
I made another attempt towards the Javascript code.

You mention that file previously: https://github.com/aeternity/testnet/blob/master/src/networking/handler.erl

I'm usually working with APIs like this: https://stripe.com/docs/api#pagination

Sample input, sample output, sample data... Here it's a little bit of a Wild West.

Previously you asked for `SHA256` - boom - done - http://plnkr.co/edit/Rqc7Yv45SUq0DuKCMQAA?p=preview

@zack-bitcoin  if you want other elements to be completed you need to invest some of your energy, it's not easy to understand the screen below:

![image](https://cloud.githubusercontent.com/assets/249703/25012629/fe263020-2068-11e7-92d7-f76af3cb7df1.png)

**Comment by zack-bitcoin** [comment]
Great, thanks for the support @GroundedSage 
@stefek99 sure, sample input and output sounds like a good idea.

**Comment by zack-bitcoin** [comment]
@stefek99 
It is a good start, but the hash still isn't equivalent to what erlang produces.
2> testnet_hasher:bin_to_hex(testnet_hasher:doit(<<>>)).
["B","6","1","3","6","7","9","A","0","8","1","4","D","9",
 "E","C","7","7","2","F","9","5","D","7"]
The hash of the empty binary should start with B6 when written as hexidecimal

**Comment by zack-bitcoin** [comment]
https://github.com/aeternity/testnet/blob/experimental/docs/api_examples.md

@stefek99  here is an example. Not complete coverage of api yet.

**Comment by stefek99** [comment]
Hash of the empty string: http://www.xorbin.com/tools/sha256-hash-calculator

![image](https://cloud.githubusercontent.com/assets/249703/25042776/8a3bc332-2112-11e7-9944-ff56e644577c.png)

Hash of the empty binary and hash of the empty string - are these any different?

-----

Cannot run the command provided by you...

`testnet_hasher:bin_to_hex(testnet_hasher:doit(<<>>)).`

...it's crashing on my machine: https://github.com/aeternity/testnet/issues/40

**Comment by zack-bitcoin** [comment]
Yes, I see now that the hash function in the testnet library was improperly set up.
I can fix it and it will generate the same outcome as the javascript library you provided.

This will make the old testnet blocks invalid.

**Comment by zack-bitcoin** [comment]
This version of the code has the best headers https://github.com/BumblebeeBat/plasmodial

**Comment by JasoonS** [comment]
I would also like to get involved here, I have spent some time working out the api from a javascript perspective but haven't gotten very far.
@stefek99 Are you still working on this? Maybe we can chat and work out how we can collaborate on this in the future? I'll send you an email now.

**Comment by stefek99** [comment]
@JasoonS I saw the email:

> would you like to do a hangouts or something

So far I didn't make much progress because I've been "busy" reporting various issues.

I suggest we get @zack-bitcoin on the call and arrange weekly dev hangouts...

**Comment by zack-bitcoin** [comment]
How about we meet weekly on the gitter? What time is best for you guys?
I am in Germany, which is in Central European Summer time zone right now.
 11 AM tomorrow is good for me.

**Comment by GroundedSage** [comment]
I am in Australia(AEST), UTC+10:00 . Anytime generally during the day and to late night I am good.
I am getting quite busy at present so when deciding on a time make the preference to others over me. I should be much more free again soon.

**Comment by andreystehno** [comment]
@zack-bitcoin I am nodejs expect. Would like to port your blockchain into nodejs. Installed your erlang server on my mac without issues. Please help me to understand where to start. How are you going to configure the communication layer between erlang server and nodejs server? via REST?

**Comment by stefek99** [comment]
I think there are some good reasons why Erlang is used: https://en.wikipedia.org/wiki/Erlang_(programming_language)

**Comment by zack-bitcoin** [comment]
Hi @andreystehno 
We are in need of your help, but we don't need the blockchain rewritten to nodejs. 
We just want a light-node that can verify block headers and merkel proofs of the consensus state.
We also want all the channel stuff re-written to javascript, so that you can do smart contracts from the browser light-node.

**Comment by GroundedSage** [comment]
Hey @zack-bitcoin I'm wondering what your thoughts are on the future with respect to client side. There is a lot of weight behind Web Assembly so I believe it will be upon us much sooner than the usual technology updates. I'm not sure how much is involved/needed for front end... as I am still new to all of this. 

Obviously its important to have something working *now* but it may not be worth investing too heavily in the client side? There is also a lot of great compile to JS languages around such as Elm, Purescript, Clojurescript etc.

**Comment by zack-bitcoin** [comment]
@GroundedSage 

The javascript goals right now are:
1) we need to be able to verify the longest chain of block headers.
2) we need to be able to verify merkel proofs.
3) Keep a record of the current channel state.
4) sign transactions.

I don't see what web assembly would be used for.

It doesn't matter to me if a compile to javascript language is used. I have heard good things about immutable.js

**Comment by askucher** [comment]
@zack-bitcoin Can we talk ? What chat do you use?

**Comment by GroundedSage** [comment]
@zack-bitcoin cool well I would be more than happy to have a serious go at this if compile to JS languages are fine. I have a lot of time on my hands at the moment though I'm still quite new to the Blockchain space and I'm not quite sure where to go to learn about the things you mentioned. As I'm not sure if another source of information will really help me here. I thoroughly going through the wiki and will start analysing the source to see if I can figure it all out.

Are you still doing those meetings weekly on Gitter? If you just set a regular time I'll block it out in my diary.

**Comment by zack-bitcoin** [comment]
Hi @askucher 
Lets chat here: https://gitter.im/BumblebeeBat/plasmodial

@GroundedSage 
If you want to learn, I recommend looking at the /docs folder.
